### PR TITLE
Authentication project implementation

### PR DESCRIPTION
**Pull request by Salvador-ON**
- It has two model: User, Post
- The User has 3 attributes:name, email, and password with the respective validation, remeber token.
- The Post has 2 attributes:comment, user_id
- In the front-end we can create new users and create new posts, when the user isn't logged in. The autor name is hidden

### CODE REVIEW

File: app/helpers/sessions_helper.rb [hunk 1]
```diff
@@ -0,0 +1,41 @@
+module SessionsHelper
+  # Logs in the given user.
+  def log_in(user)
+    session[:user_id] = user.id
+  end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: Not needed

File: app/helpers/sessions_helper.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  # Remembers a user in a persistent session.
+  def remember(user)
+    user.remember
+    cookies.permanent.signed[:user_id] = user.id
+    cookies.permanent[:remember_token] = user.remember_token
+  end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb:6]: You should set the `@current_user` to the one who signed

File: app/helpers/sessions_helper.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    if (user_id = session[:user_id])
+      @current_user ||= User.find_by(id: user_id)
+    elsif (user_id = cookies.signed[:user_id])
+      user = User.find_by(id: user_id)
+      if user&.authenticated?(cookies[:remember_token])
+        log_in user
+        @current_user = user
+      end
+    end
+  end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: Too long!!!
Re-read this [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/current_user_requirement.png)

File: app/helpers/sessions_helper.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+  def current_user=(user)
+    user == current_user
+  end
```
Comment (maelfosso) [inline app/helpers/sessions_helper.rb]: It's not `user == current_user`
You should defined the `@current_user` here

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required  :red_circle: 

Hello, thank you the job done! Really simple!!
However, there are changes to do

1. After login, I check the database
```ruby
2.6.3 :001 > User.find_by(:email => "alice@bar.com")
  User Load (0.2ms)  SELECT  "users".* FROM "users" WHERE "users"."email" = ? LIMIT ?  [["email", "alice@bar.com"], ["LIMIT", 1]]
 => #<User id: 12, name: "Alice", email: "alice@bar.com", password_digest: "$2a$12$N2j/9gnAZ3a9QC/tpy/eeu.22MWOiq7FK/DpgKynJjD...", member_status: nil, created_at: "2020-03-17 22:54:05", updated_at: "2020-03-17 22:54:05", remember_token: nil> 
2.6.3 :002 > 
```
You could notice that the `remember_token` is nil. It should have the token as indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/sign_in_method_requirement.png)

2. Forgot about **using SESSION**. 
> The `remember_me` token the project describes consists of storing a hash in a cookie and using it to determine in a user is logged in. Do not confuse it with the Remember me [ ] checkbox that many websites show. That is a more complex solution, which is not required here

3. When **log_out**, you should delete the `remember_token` from the cookies and set it to `nil` into the user.

4. At the end of the sign in, you should set the current user variable ie `@current_user`.

5. Read the comments below, please.

Kindly make the suggested changes and submit a new code review request.
Thanks

Mael FOSSO

**Comment by andresporras3423** [comment]
Thanks for the feedback

* not using session anymore
* login with cookies
* cookies=nil when log out
* using current_user_equal  method to update current user
* deleted unnecessary methods

File: app/models/user.rb [hunk 5]
```diff
@@ -17,19 +17,19 @@ def self.digest(string)
...
-  end
+  # def remember
+  #   self.remember_token = User.new_token
+  #   update_attribute(:remember_token, User.digest(remember_token))
+  # end
 
   # Returns true if the given token matches the digest.
   def authenticated?(remember_token)
-    BCrypt::Password.new(remember_token).is_password?(remember_token)
+    self.remember_token == remember_token
```
Comment (samgaco) [inline app/models/user.rb:21]: :+1:

**Review by samgaco — COMMENTED** [review]
Hi @andresporras3423 and @Salvador-ON :high_brightness: 

You have implemented the changes, the project is approved :+1: 

-----------------

* ###### STATUS OF THE PROJECT: APPROVED :heavy_check_mark:

**Comment by andresporras3423** [comment]
Thank you

**Pull request merged**